src/devices/machine/spifi3.h

```diff
@@ -0,0 +1,305 @@
+			return removed_value;
+		}
+	};
+
+	// State tracking variables
+	dma_direction dma_dir;
+	scsi_mode mode;
+	scsi_data_target xfr_data_source;
+	uint32_t state;
+	uint32_t xfr_phase;
+	uint32_t command_pos;
+	bool irq = false;
+	bool drq = false;
+	uint32_t tcounter;
+	uint8_t sync_period = 5; // TODO: appropriate value for SPIFI
+	uint8_t clock_conv = 2; // TODO: appropriate value for SPIFI
+	uint32_t bus_id;
+	spifi_queue<uint8_t> m_even_fifo;
+	spifi_queue<uint8_t> m_odd_fifo;
+	emu_timer *tm;
+
+	// I/O ports
+	devcb_write_line m_irq_handler;
+	devcb_write_line m_drq_handler;
+
+	// State-related methods
+	TIMER_CALLBACK_MEMBER(tick);
+	void step(bool timeout);
+	void check_irq();
+	void check_drq();
+	void reset_disconnect();
+	void send_byte(scsi_data_target data_source);
+	void recv_byte();
+	void function_complete();
+	void function_bus_complete();
+	void bus_complete();
+	void dma_set(dma_direction dir);
+	void decrement_tcounter(uint32_t count = 1);
+	bool transfer_count_zero();
+	void delay(uint32_t cycles);
+	void delay_cycles(uint32_t cycles);
+	void arbitrate();
+	void clear_fifo();
+	void auto_phase_transfer(uint32_t new_phase);
+	void start_autodata(uint32_t data_phase);
+	void start_autostat();
+	void start_automsg(uint32_t msg_phase);
+	void start_autocmd();
+	dma_direction dma_setting(uint32_t target_id);
+
+	// Register processing methods
+	uint32_t get_target_id();
+	bool autodata_active(uint32_t target_id);
+	bool autodata_in(uint32_t target_id);
+	bool autodata_out(uint32_t target_id);
+	void autostat_done(uint32_t target_id);
+	bool autostat_active(uint32_t target_id);
+	bool automsg_active();
+	bool autocmd_active();
+
+	// Register accessors
+	uint32_t spstat_r();
+	uint32_t cmlen_r();
+	void cmlen_w(uint32_t data);
+	uint32_t cmdpage_r();
+	void cmdpage_w(uint32_t data);
+	uint32_t count_r(offs_t offset);
+	void count_w(offs_t offset, uint32_t data);
+	uint32_t intr_r();
+	void intr_w(uint32_t data);
+	uint32_t imask_r();
+	void imask_w(uint32_t data);
+	uint32_t prstat_r();
+	uint32_t init_status_r();
+	uint32_t fifoctrl_r();
+	void fifoctrl_w(uint32_t data);
+	uint32_t data_xfer_r();
+	void data_xfer_w(uint32_t data);
+	uint32_t autocmd_r();
+	void autocmd_w(uint32_t data);
+	uint32_t autostat_r();
+	void autostat_w(uint32_t data);
+	uint32_t select_r();
+	void select_w(uint32_t data);
+	void prcmd_w(uint32_t data);
+	uint32_t auxctrl_r();
+	void auxctrl_w(uint32_t data);
+	uint32_t autodata_r();
+	void autodata_w(uint32_t data);
+	uint32_t identify_r();
+	void identify_w(uint32_t data);
+	uint32_t scsi_status_r();
+	void scsi_status_w(uint32_t data);
+	uint32_t icond_r();
+	void icond_w(uint32_t data);
+	uint32_t fastwide_r();
+	void fastwide_w(uint32_t data);
+	uint32_t exctrl_r();
+	void exctrl_w(uint32_t data);
+
+	// Command buffer accessors
+	uint8_t cmd_buf_r(offs_t offset);
+	void cmd_buf_w(offs_t offset, uint8_t data);
+
+	// Data helpers
+	inline bool dma_command(dma_direction current_direction)
```

Comment:
There’s no point using `inline` here – it’s implied when a member function body is included with the declaration.  It should be `const`-qualified, though.